### ISSUE TITLE
Sound "buzz" when playing video while in a call

### ISSUE THREAD

**Issue by ruipfcosta** [QuickBlox/quickblox-ios-sdk#257]
I am integrating QuickBlox in an app where the user will have the ability to minimize the call and continue to interact with the app as normal. This app has the ability to display some videos, and when I'm in a call and and the app displays a video, a "buzzing" sounds starts (like if something was broken) and it can only be stopped if I terminate the call. The receiver stops getting any sound as soon as this "buzz" starts on the caller.
Is there any solution for this problem?

**Comment by Pro100Andrey** [comment]
Hi @ruipfcosta !! Do you use simulator ?

**Comment by ruipfcosta** [comment]
No, I'm using 2 iPads.

**Comment by Pro100Andrey** [comment]
@ruipfcosta  Please set other audioCodec:
`
[QBRTCConfig mediaStreamConfiguration].audioCodec = QBRTCAudioCodecISAC;
`

default - QBRTCAudioCodecOpus

**Comment by ruipfcosta** [comment]
I tried all available audio codecs (QBRTCAudioCodecOpus, QBRTCAudioCodecISAC or QBRTCAudioCodeciLBC) but the result is still the same. Any ideas?

**Comment by Pro100Andrey** [comment]
What is your cpu load on both devices?

**Comment by ruipfcosta** [comment]
I'm using an iPad 2 (iOS 9.2) to initiate the call (the CPU load gets to ~120%). The CPU load on the receiver (iPad Mini) is lower, but the incoming call interface is different which explains it.

You can find the logs for the call here:
http://pastebin.com/raw/NS3z8DM3

**Comment by ruipfcosta** [comment]
Ok this might be important information, I tested multiple devices and so far the only one where the sound "buzz" does not occur is on iOS 8.3. All the other devices affected by this issue are on iOS 9.x.

**Comment by andrewjackman** [comment]
I have this issue too. I'm on the following versions:
```
pod 'QuickBlox', '~> 2.6.3'
pod 'Quickblox-WebRTC', '~> 2.1.1'
```
Also running iOS 9.x on all devices. I have tested on iOS 8.x as I don't have any devices running it.

**Comment by Raikerian** [comment]
What camera resolutions and video formats are you using? What is your FPS during call?
Any other settings that can help us?

**Comment by ruipfcosta** [comment]
@Raikerian I posted some logs on a previous comment, not sure if this line helps:

```rtc::Best device format is: <AVCaptureDeviceFormat: 0x17457f80 'vide'/'420f'  640x 480, { 4- 30 fps}, fov:54.910, ISO:135.0-2160.0, SS:0.000061-0.250000>```

**Comment by Raikerian** [comment]
Could you please try lower resolution and fps for your video call until cpu load will be lower then 100% for old devices and tell us the result? And use audio codec QBRTCAudioCodecISAC.

**Comment by ruipfcosta** [comment]
Hi @Raikerian @iSevenDays , 

Below you can find a modified WebRTC sample where we can see the sound buzz issue happening, but not the low volume audio one (https://github.com/QuickBlox/quickblox-ios-sdk/issues/259). I made two modifications:

* Pressing the "dynamic" button during a call plays a video;
* Pressing the screen share button during a call plays an audio file.

It looks like the sound buzz issue is caused by the video format used (M4V), we tried with different M4V files and happens every time. However, if we play an MP4 file instead everything is fine.

[sample-videochat-webrtc-modified.zip](https://github.com/QuickBlox/quickblox-ios-sdk/files/104885/sample-videochat-webrtc-modified.zip)

**Comment by Raikerian** [comment]
Hi @ruipfcosta 
we weren't able to reproduce "buzz" sound in your sample. Please provide us extended steps on how to reproduce it.

**Comment by Raikerian** [comment]
Actually we were able to catch that "buzz" sound, but it was fixed by lower resolution and QBRTCAudioCodecISAC for old iPad devices, have you tried it yet?
You can set in the samples settings and test it out.

**Comment by ruipfcosta** [comment]
So I'm testing the modified sample using three iPads:

* iPad Mini iOS 8.3
* iPad Mini iOS 9.1
* iPad Air iOS 9.2.1

On the devices with iOS 9.x the "buzz" sound is happening, no matter what settings I use - I tried all the possible audio codecs, renderer types, lowest resolution/bandwidth/framerate.
On the device with iOS 8 the "buzz" sound does not happen, but the video sound is really low (I almost can't hear it, but it's definitely playing). I also tried multiple settings combinations without success.
What devices/iOS version are you using?

**Comment by Raikerian** [comment]
We have tested it on iPad 2 (os 8) and iPad 3 (os 9) and there is no such sound. Could you maybe provide an example of this sound?
Also what is your cpu load when this sound is reproducing?

**Comment by ruipfcosta** [comment]
The CPU load on the sample is floating between 40-80% when that happens. I recorded a small video where you can see the behaviour we are experiencing.

 [buzz.zip](https://github.com/QuickBlox/quickblox-ios-sdk/files/105126/buzz.zip)

**Comment by Raikerian** [comment]
Is that sample you sent to us on video? Cause sample you provided contains only music.

**Comment by Raikerian** [comment]
Please try resetting location and privacy in settings (it should also reset microphone settings).

**Comment by ruipfcosta** [comment]
Yes, I confirmed and it's the same sample I shared in a previous post here. One buttons plays a video (sample.m4v), and another button plays a sound file (example.wav) like I mentioned.
Reseting location and privacy didn't have any impact on the problem.

**Comment by Raikerian** [comment]
We are currently investigating this problem and will post update soon.

**Comment by Raikerian** [comment]
Can you try QBRTCVideoCodecH264 codec on both devices and tell us if there any results?

**Comment by ruipfcosta** [comment]
Just tested with H264 video and the problem persists.

**Comment by Raikerian** [comment]
Investigation is in progress due to us not being able to reproduce it.

**Comment by ruipfcosta** [comment]
Hi guys, did you have any success replicating this issue using the sample is shared  [before](https://github.com/QuickBlox/quickblox-ios-sdk/issues/257#issuecomment-175004422)? Is there anything I can do to help you replicate the issue?

Thanks!

**Comment by iSevenDays** [comment]
Hi ruipfcosta,

We have found one possible cause of that error.

Please try to disable QBRTCSoundRouter, do not use initialize/deinitialize and see if error happens again.

If your problem is related to sound router, then we will fix it.

Thanks!

**Comment by ruipfcosta** [comment]
Hi,

The problem is still happening after removing the QBRTCSoundRouter init/deinit.

Thanks.

**Comment by ruipfcosta** [comment]
I also tested on the modified WebRTC sample and also happens there.

**Comment by iSevenDays** [comment]
Hi ruipfcosta,
I've found another possible problem.

I've faced the "buzz" sound when I do manage [AVAudioSession sharedSession] by myself, e.g. calling setActive:. 

Do you manage [AVAudioSession sharedSession]?

**Comment by ruipfcosta** [comment]
Hi @iSevenDays, unfortunately this had no effect on the problem, as I removed all the code where we are dealing with AVAudioSession and the buzz persists.
I would like to take a step back and see if there is something we can do to help you replicate the problem with the sample I provided. What is the behaviour you see when you press the button I indicated during a call? Do you see the sample video playing normally without any sound issues?
Also, I'm happy to take this discussion to other channel (i.e. Skype) if you feel like it can be more helpful.

Thanks!

**Comment by iSevenDays** [comment]
Hi ruipfcosta,

Thank you for being patient!

We still do investigate the issue. Also I've faced this sound with old webrtc revision, but I can not reproduce the issue in the latest webrtc revision. They have added improvements for low-performance devices (with 1-core CPU) and also they have changed audio session manager.

We will let you know once we fix it.

Thanks!

**Comment by Raikerian** [comment]
Hi again @ruipfcosta @andrewjackman 

We have just released our new WebRTC SDK 2.3 which is containing completely new audio module.
You can now configure category, category options and mode of AVAudioSession used in WebRTC, read more here: http://quickblox.com/developers/Sample-webrtc-ios#Audio_Session_.28Previously_Sound_Router.29

Category options provides such things as Ducking and Mixing, you might want to look more into it.
Here is full changelog of version 2.3: http://quickblox.com/developers/Sample-webrtc-ios#Framework_changelog

Anyway, if you will have any other problems or questions, feel free to start new issue to let us know. 

Have a nice holidays.
